Mathlib/CategoryTheory/Sites/RegularExtensive.lean

```diff
@@ -0,0 +1,131 @@
+epimorphism.
+
+The second one is called the *extensive* coverage and for that to exist, the category `C` must
+satisfy a condition called `Extensive C`. This means `C` has finite coproducts and that those
+are preserved by pullbacks. The covering sieves of this coverage are generated by presieves
+consisting finitely many arrows that together induce an isomorphism from the coproduct to the
+target.
+
+In `extensive_union_regular_generates_coherent`, we prove that the union of these two coverages
+generates the coherent topology on `C` if `C` is precoherent, extensive and regular.
+
+TODO: figure out under what conditions `Regular` and `Extensive` are implied by `Precoherent` and
+vice versa.
+
+-/
+
+universe v u w
+
+namespace CategoryTheory
+
+open Limits
+
+variable (C : Type u) [Category.{v} C]
+
+/--
+The condition `Regular C` is property that effective epis can be "pulled back" along any
+morphism. This is satisfied e.g. by categories that have pullbacks that preserve effective
+epimorphisms (like `Profinite` and `CompHaus`), and categories where every object is projective
+(like  `Stonean`).
+-/
+class Regular : Prop where
+  /--
+  For `X`, `Y`, `Z`, `f`, `g` like in the diagram, where `g` is an effective epi, there exists
+  an object `W`, a morphism `i : W ⟶ Z` and an effective epi `h : W ⟶ X` making the diagram
+  commute.
+  ```
+  W --i-→ Z
+  |       |
+  h       g
+  ↓       ↓
+  X --f-→ Y
+  ```
+  -/
+  exists_fac : ∀ {X Y Z : C} (f : X ⟶ Y) (g : Z ⟶ Y) [EffectiveEpi g],
+    (∃ (W : C) (h : W ⟶ X) (_ : EffectiveEpi h) (i : W ⟶ Z), i ≫ g = h ≫ f)
+
+/--
+Describes the property of having pullbacks of morphsims into a finite coproduct, where one
+of the morphisms is an inclusion map into the coproduct (up to isomorphism).
+-/
+class HasPullbacksOfInclusions : Prop where
+    /-- For any morphism `f : X ⟶ Z`, where `Z` is the coproduct of `i : (a : α) → Y a ⟶ Z` with
+    `α` finite, the pullback of `f` and `i a` exists for every `a : α`. -/
+    has_pullback : ∀ {X Z : C} {α : Type w} (f : X ⟶ Z) {Y : (a : α) → C}
+    (i : (a : α) → Y a ⟶ Z) [Fintype α] [HasCoproduct Y] [IsIso (Sigma.desc i)] (a : α),
+    HasPullback f (i a)
+
+instance [HasPullbacksOfInclusions C] {X Z : C} {α : Type w} (f : X ⟶ Z) {Y : (a : α) → C}
+    (i : (a : α) → Y a ⟶ Z) [Fintype α] [HasCoproduct Y] [IsIso (Sigma.desc i)] (a : α) :
+    HasPullback f (i a) := HasPullbacksOfInclusions.has_pullback f i a
+
+/--
+If `C` has pullbacks then it has the pullbacks relevant to `HasPullbacksOfInclusions`.
+-/
+instance (priority := 10) [HasPullbacks C] :
+  HasPullbacksOfInclusions C := ⟨fun _ _ _ => inferInstance⟩
+
+/--
+A category is *extensive* if it has all finite coproducts and those coproducts are preserved
+by pullbacks (we only require the relevant pullbacks to exist, via `HasPullbacksOfInclusions`).
+-/
+class Extensive extends HasFiniteCoproducts C, HasPullbacksOfInclusions C : Prop where
```

Comment:
Also this is not standard terminology I think. And we already have [CategoryTheory.FinitaryExtensive](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Extensive.html#CategoryTheory.FinitaryExtensive). We probably should figure out a different name, or try to relate the two notions.

Reply:
OK, so maybe let's add a note saying that sooner or later we should relate the two notions.

Reply:
I'll do that. It seems like if we can prove that our notion implies that finite coproducts are "van Kampen", then we could replace the existing `FinitaryExtensive` by our notion, or at least prove that our notion implies `FinitaryExtensive`. But for now let's keep both and just add the note. What about the name? `FinitaryExtensive'`?

Reply:
Let's keep `Extensive`, I have the impression `FinitaryExtensive` is almost not used in mathlib.